Include change history in Publishing API payload

brucebolt opened the pull request:
Now that we store a full `change_history` for an edition, we need to send this full history (along with the `first_published_date` and `public_updated_at`) to Publishing API.

For the first edition, we always sent a fixed change note.  For subsequent major editions, the user entered change note is used.

This change also moves `Payload` out of the `PreviewDraftEditionService` namespace, into `PublishingApiPayload` since this class is used in multiple contexts, not just when dealing with the preview of draft editions.

Not to be merged until #1806 is merged and deployed.

Trello card: https://trello.com/c/KBFWEkKG

kevindew commented on spec/lib/publishing_api_payload_spec.rb:
```diff
@@ -18,6 +18,70 @@
+    end
+
+    it "includes a first_published_at and public_updated_at timestamp for published editions" do
+      freeze_time do
```
I wonder if you can avoid the need for freeze time here by setting fixed dates? I assume we're more testing it passes dates on rather than it is the current time.

brucebolt replied:
I've removed the freeze_time from all the tests apart from those where we expect the current time

brucebolt commented on lib/publishing_api_payload.rb:
```diff
@@ -22,7 +22,9 @@ def payload
       auth_bypass_ids: auth_bypass_ids,
+      public_updated_at: history.public_updated_at,
     }
+    payload[:first_published_at] = history.first_published_at if edition.published?
```
A nil value is not valid (according to govuk-content-schemas) and an unrelated test fails due to an invalid schema without this guard clause.

kevindew replied:
Oh bummer hmm this should probably check that history.first_published_at is set then since we'd want to set this when an edition is backdated.

brucebolt commented on spec/factories/metadata_revision_factory.rb:
```diff
@@ -2,6 +2,7 @@
   factory :metadata_revision do
     update_type { "major" }
     document_type_id { build(:document_type, path_prefix: "/prefix").id }
+    change_history { [] }
```
Have removed this

brucebolt commented on spec/lib/publishing_api_payload_spec.rb:
```diff
@@ -18,6 +18,70 @@
+        }
+
+        expect(payload).to match a_hash_including(payload_hash)
+        expect(payload).not_to include(:first_published_at)
```
Removed this, we're already testing it elsewhere

brucebolt commented on spec/lib/publishing_api_payload_spec.rb:
```diff
@@ -18,6 +18,70 @@
+    it "includes a public_updated_at timestamp for draft editions" do
+      freeze_time do
+        document_type = build(:document_type)
+        edition = build(:edition, document_type: document_type)
```
Yes, we don't need a document_type, so I've removed that

brucebolt commented:
Thanks for the review @kevindew.  I've moved the History class out into a namespaced file and added several unit tests for this class that cover multiple scenarios.

kevindew commented on spec/lib/publishing_api_payload/history_spec.rb:
```diff
@@ -0,0 +1,110 @@
+  end
+
+  describe "#change_history" do
+    let(:edition) { build(:edition, first_published_at: "2020-02-22 11:00:00") }
```
Might be clearer to name this as first_edition

Might also be useful to have a let of `initial_change_note` since you reference the same one a few times in tests.

brucebolt replied:
Good points.  I've changed the name of `edition` to `first_edition` and added `initial_change_note`.

kevindew commented on spec/lib/publishing_api_payload/history_spec.rb:
```diff
@@ -0,0 +1,110 @@
+        public_timestamp: "2020-02-22 11:00:00",
+      }
+
+      expect(described_class.new(second_edition).change_history.first).to eq(expected_change_note)
```
I wasn't sure if this test quite captured the scenario as it only tests one of the change notes.

Might be better if the assertion was more along the lines of:

`expect(described_class.new(second_edition).change_history).to eq([expected_change_note])`

brucebolt replied:
Made that change, and also made a similar change for other tests where we expect an array to be returned by `change_history`.

kevindew commented on spec/lib/publishing_api_payload/history_spec.rb:
```diff
@@ -0,0 +1,110 @@
+      ]
+
+      expect(described_class.new(second_edition).change_history).to eq(expected_change_notes)
+    end
```
Be useful to also have a test that change notes set before a backdating aren't shown.

brucebolt replied:
Added a test for this

kevindew commented on spec/lib/publishing_api_payload_spec.rb:
```diff
@@ -18,6 +18,62 @@
+        ],
+      }
+
+      expect(payload[:details]).to match a_hash_including(expected_details_hash)
```
I don't think you need to test these bits here since we're covering a bunch of this in the history object test.

We can probably do something like:

```
it "delegates to PublishingApiPayload::History to populate change_history" do
  history = instance_double(
    PublishingApiPayload::History, 
    change_history: [{ note: "note", public_timestamp: Time.current }],
    public_updated_at: Time.current,
    first_published_at: Time.current,
  )
  allow(PublishingApiPayload::History).to receive(:new).and_return(history)
  payload = described_class.new(build(:edition)).payload
  expect(payload[:details][:change_history]).to match(history.change_history)
end
```

brucebolt replied:
Sure.  I've replaced the test with the one you've proposed, and ensured the `PublishingApiPayload::History` spec covers the situations we were previously testing here.

brucebolt commented:
Thanks for the review @kevindew.  I've made all the proposed minor changes.
